File: frontend/src/api/recall.ts
// ============================================================
// Recall API（对接 marketing-service /api/marketing/recall）
// P5-B90 随访 Flow：复诊提醒列表/新建/提醒/确认/预约/改期/跳过。
// axios client 拦截器不透传 .data，消费侧须 res.data（B86 教训）。
// 词表与 V48 chk 一致：source=DOCTOR_ADVICE/COURSE_FOLLOW/SYSTEM_AUTO/
// MANUAL，method=PHONE/WECHAT/SMS/IN_STORE，status 五态；
// 状态机服务端权威，非法转移 409（errMsg 透传 message 中文文案）。
// ============================================================
import client from './client'

/**
 * 复诊提醒行（GET /marketing/recall 真实字段）。
 * timeline 为 JSON 数组字符串（元素 {at, by, action, detail?}），前端 store 适配层 JSON.parse。
 * 日期字段 yyyy-MM-dd；notifiedAt/createdAt 为 +8 业务时区 ISO 时间串。
 */
export interface RecallRow {
  id: string
  recallNo: string
  customerId: string
  customerName: string
  source: string
  reason: string
  relatedEmrNo: string | null
  relatedOrderNo: string | null
  lastVisitDate: string | null
  dueDate: string
  method: string
  status: string
  notifiedByName: string | null
  notifiedAt: string | null
  customerReply: string | null
  confirmedDate: string | null
  skipReason: string | null
  note: string | null
  timeline: string
  createdAt: string
  /** 来源规则编号：非空 = Flow 引擎自动创建（页面展示「自动」徽标） */
  ruleNo: string | null
  storeCode: string | null
}

/** KPI 四键（随列表响应返回；前端 store 按本地 computed 口径自算，不消费此对象） */
export interface RecallKpi {
  overdue: number
  todayPending: number
  upcoming: number
  conversionRate: number
}

export interface RecallListResp {
  recalls: RecallRow[]
  kpi: RecallKpi
}

/** 新建命令（source 四来源；dueDate=yyyy-MM-dd 必填；method 默认 PHONE；customerName 后端按 customerId 解析） */
export interface ScheduleRecallCmd {
  customerId: string
  source: string
  reason: string
  relatedEmrNo?: string
  relatedOrderNo?: string
  lastVisitDate?: string
  dueDate: string
  method?: string
  note?: string
}

/** 列表（status 精确五态；kw 模糊姓名/事由/单号；storeCode 精确）＋全量 KPI */
export const fetchRecalls = (params?: { status?: string; kw?: string; storeCode?: string }) =>
  client.get<RecallListResp>('/marketing/recall', { params })

/** 新建复诊提醒（须 recall:create；落 PENDING；客户硬校验） */
export const scheduleRecall = (body: ScheduleRecallCmd) =>
  client.post<RecallRow>('/marketing/recall', body)

/** 执行提醒 PENDING → NOTIFIED（须 recall:edit） */
export const notifyRecall = (recallNo: string, method?: string) =>
  client.post<RecallRow>(`/marketing/recall/${recallNo}/notify`, { method })

/** 登记客户确认 NOTIFIED → CONFIRMED（confirmedDate=yyyy-MM-dd） */
export const confirmRecall = (recallNo: string, body: { reply?: string; confirmedDate?: string }) =>
  client.post<RecallRow>(`/marketing/recall/${recallNo}/confirm`, body)

/** 标记已预约 NOTIFIED/CONFIRMED → BOOKED */
export const bookRecall = (recallNo: string, detail?: string) =>
  client.post<RecallRow>(`/marketing/recall/${recallNo}/book`, { detail })

/** 改期（仅 PENDING/NOTIFIED；NOTIFIED → PENDING 清通知信息；dueDate=yyyy-MM-dd） */
export const rescheduleRecall = (recallNo: string, body: { dueDate: string; note?: string }) =>
  client.post<RecallRow>(`/marketing/recall/${recallNo}/reschedule`, body)

/** 跳过（reason 必填 ≤200 字；PENDING/NOTIFIED/CONFIRMED → SKIPPED） */
export const skipRecall = (recallNo: string, reason: string) =>
  client.post<RecallRow>(`/marketing/recall/${recallNo}/skip`, { reason })

File: backend/marketing-service/src/main/java/com/meiyun/marketing/Recall.java
package com.meiyun.marketing;

import jakarta.persistence.*;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;

import java.time.LocalDate;
import java.time.OffsetDateTime;

/**
 * 复诊召回（P5-B90，/recall 切真）。
 * 状态机（服务端唯一权威，非法转移 409）：
 * PENDING → [NOTIFIED, SKIPPED]；NOTIFIED → [CONFIRMED, BOOKED, SKIPPED, PENDING 改期]；
 * CONFIRMED → [BOOKED, SKIPPED]；BOOKED / SKIPPED 终态。
 * 四来源：DOCTOR_ADVICE 医嘱 / COURSE_FOLLOW 疗程跟进 / SYSTEM_AUTO 系统自动 / MANUAL 人工。
 * 四方式：PHONE 电话 / WECHAT 企微 / SMS 短信 / IN_STORE 到店。
 * timeline 为 JSONB 文本 [{at,by,action,detail}]，每次状态转移 append 一条。
 * 与 txn followup 表零共享（D5 诊疗边界）。
 */
@Entity
@Table(name = "recall")
@Getter @Setter @NoArgsConstructor
public class Recall {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Column(name = "id")
    private Long id;

    /** 召回单号：RC + yyyyMMdd + - + 6 位当日序号（BizNoGenerator）。 */
    @Column(name = "recall_no", nullable = false, unique = true, length = 32)
    private String recallNo;

    @Column(name = "customer_id", length = 16)
    private String customerId;

    /** 客户姓名冗余列：创建时经 CustomerDirectoryClient 解析落列。 */
    @Column(name = "customer_name", length = 64)
    private String customerName;

    /** DOCTOR_ADVICE / COURSE_FOLLOW / SYSTEM_AUTO / MANUAL。 */
    @Column(nullable = false, length = 16)
    private String source;

    /** 召回事由。 */
    @Column(nullable = false, length = 200)
    private String reason;

    /** 关联病历号（可空，逻辑引用不建外键）。 */
    @Column(name = "related_emr_no", length = 32)
    private String relatedEmrNo;

    /** 关联订单号（可空，逻辑引用不建外键）。 */
    @Column(name = "related_order_no", length = 32)
    private String relatedOrderNo;

    /** 末次到店 / 治疗日。 */
    @Column(name = "last_visit_date")
    private LocalDate lastVisitDate;

    /** 应召回日（超期 / 今日待提醒 / 3 日到期 KPI 锚，业务时区 +8）。 */
    @Column(name = "due_date", nullable = false)
    private LocalDate dueDate;

    /** PHONE / WECHAT / SMS / IN_STORE。 */
    @Column(nullable = false, length = 16)
    private String method;

    /** PENDING / NOTIFIED / CONFIRMED / BOOKED / SKIPPED。 */
    @Column(nullable = false, length = 16)
    private String status = "PENDING";

    /** 提醒操作人（B90 卡1 增补：前端 notifiedByName 契约直出，避免 timeline 解析）。 */
    @Column(name = "notified_by", length = 32)
    private String notifiedBy;

    /** 提醒时间（B90 卡1 增补：前端 notifiedAt 契约直出）。 */
    @Column(name = "notified_at")
    private OffsetDateTime notifiedAt;

    /** 客户回复。 */
    @Column(name = "customer_reply", length = 500)
    private String customerReply;

    /** 确认复诊日。 */
    @Column(name = "confirmed_date")
    private LocalDate confirmedDate;

    /** 跳过原因（B90 卡1 增补：前端 skipReason 契约直出）。 */
    @Column(name = "skip_reason", length = 200)
    private String skipReason;

    @Column(length = 500)
    private String note;

    /** 时间线 JSONB：[{at,by,action,detail}]，每次状态转移 append 一条。 */
    @Column(columnDefinition = "jsonb")
    private String timeline;

    /** 来源规则：逻辑引用 automation_rule.rule_no，NULL=人工创建（SYSTEM_AUTO 来源时非空）。 */
    @Column(name = "rule_no", length = 32)
    private String ruleNo;

    /** 门店码：NULL=全连锁，填值=客户归属门店（铁律-1-D）。 */
    @Column(name = "store_code", length = 32)
    private String storeCode;

    @Column(name = "created_by", length = 32)
    private String createdBy;

    @Column(name = "created_at", nullable = false)
    private OffsetDateTime createdAt = OffsetDateTime.now();

    @Column(name = "updated_at", nullable = false)
    private OffsetDateTime updatedAt = OffsetDateTime.now();
}
